toolchain/check/handle_impl.cpp

```diff
@@ -210,23 +210,41 @@ static auto PopImplIntroducerAndParamsAsNameComponent(
                               }));
   }
 
-  Parse::NodeId first_param_node_id =
-      context.node_stack().PopForSoloNodeId<Parse::NodeKind::ImplIntroducer>();
+  Parse::Tree::PostorderIterator first_param_iter(
+      context.node_stack().PopForSoloNodeId<Parse::NodeKind::ImplIntroducer>());
   // Subtracting 1 since we don't want to include the final `{` or `;` of the
   // declaration when performing syntactic match.
-  // TODO: Following proposal #3763, we should exclude any `where` clause, and
-  // add `Self` before `as` if needed, see:
-  // https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p3763.md#redeclarations
   auto node_kind = context.parse_tree().node_kind(end_of_decl_node_id);
   CARBON_CHECK(node_kind == Parse::NodeKind::ImplDefinitionStart ||
                node_kind == Parse::NodeKind::ImplDecl);
-  Parse::NodeId last_param_node_id(end_of_decl_node_id.index - 1);
+  Parse::Tree::PostorderIterator last_param_iter(end_of_decl_node_id);
+  --last_param_iter;
+  // Following proposal #3763, exclude a final `where` clause, if present. See:
+  // https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p3763.md#redeclarations
+  node_kind = context.parse_tree().node_kind(*last_param_iter);
+  if (node_kind == Parse::NodeKind::WhereExpr) {
+    int where_operands_to_skip = 1;
+    --last_param_iter;
+    CARBON_CHECK(first_param_iter < last_param_iter);
+    do {
+      node_kind = context.parse_tree().node_kind(*last_param_iter);
+      if (node_kind == Parse::NodeKind::WhereExpr) {
+        // If we have a nested `where`, we need to see another `WhereOperand`
+        // before we find the one that matches our original `WhereExpr` node.
+        ++where_operands_to_skip;
+      } else if (node_kind == Parse::NodeKind::WhereOperand) {
+        --where_operands_to_skip;
+      }
+      --last_param_iter;
+      CARBON_CHECK(first_param_iter < last_param_iter);
+    } while (where_operands_to_skip > 0);
+  }
```

Comment:
Would it be better to defer doing this until we build the `DeclParams` objects in `MergeImplRedecl`? That'd mean we don't need to do the scan here until we're about to compare two token sequences anyway, which would avoid doing this scan entirely in cases where the impl isn't in the same bucket as any other impl (which is probably a common case), and would likely increase locality given that we'd be doing two checks of nearby tokens at around the same time.

Reply:
@jonmeow What do you think?

Reply:
I agree that we could avoid the scan by doing saving something in the `WhereExpr` handling, but it seems really fragile. There are a lot of places where a `where` clause can occur in an `impl` declaration that we don't want to truncate at.
```
impl forall [T:! I where...] ...
impl DynPtr(Container where .ElementType = i32) as ...
impl ... as J(K where ...)
impl ... as (... where ...) where ...
```
I think the current code is easier to get right since it directly looks in the one place we want to operate on.

Reply:
In each of the cases you're showing, it seems like there'd be something on the NodeStack indicating the nesting:

- `[`
- `(`
- `(`
- `(`

Maybe it's worth thinking about further?